src/economy/waterway.cc

```diff
@@ -157,57 +153,67 @@ void Waterway::postsplit(Game& game, Flag& flag) {
 
 	// change waterway size and reattach
 	flags_[FlagEnd] = &flag;
-	set_path(game, path);
+	set_path(egbase, path);
 
 	const Direction dir = get_reverse_dir(path_[path_.get_nsteps() - 1]);
 	flags_[FlagEnd]->attach_road(dir, this);
 	flagidx_[FlagEnd] = dir;
 
 	// recreate waterway markings
-	mark_map(game);
+	mark_map(egbase);
 
 	// create the new waterway
 	Waterway& newww = *new Waterway();
 	newww.set_owner(get_owner());
 	newww.flags_[FlagStart] = &flag;  //  flagidx will be set on init()
 	newww.flags_[FlagEnd] = &oldend;
-	newww.set_path(game, secondpath);
+	newww.set_path(egbase, secondpath);
 
 	// Initialize the new waterway
-	newww.init(game);
+	newww.init(egbase);
 
-	if (ferry_) {
-		assert(ferry_->get_location(game) == this);
-		// We assign the ferry to the waterway part it currently is on
-		bool other = true;
-		const Coords pos = ferry_->get_position();
-		Coords temp(flags_[FlagStart]->get_position());
-		for (uint32_t i = 0; i < path_.get_nsteps(); i++) {
-			if (temp == pos) {
-				other = false;
-				break;
+	if (upcast(Game, game, &egbase)) {
+		if (ferry_) {
```

Comment:
Move this line above the `upcast`, then you will only have an `else` below

Reply:
The `if(upcast(Game` block also contains some lines that will be run regardless whether `ferry_` is null or not, so we'd still need an upcast in any case and wouldn't win anything by changing the order.

Reply:
You're right, I overlooked that.